src/relations/spark_sa.py

```diff
@@ -0,0 +1,333 @@
+    old_data = json.loads(event.relation.data[bucket].get("data", "{}"))
+    # Retrieve the new data from the event relation databag.
+    new_data = (
+        {key: value for key, value in event.relation.data[event.app].items() if key != "data"}
+        if event.app
+        else {}
+    )
+
+    # These are the keys that were added to the databag and triggered this event.
+    added = new_data.keys() - old_data.keys()
+    # These are the keys that were removed from the databag and triggered this event.
+    deleted = old_data.keys() - new_data.keys()
+    # These are the keys that already existed in the databag,
+    # but had their values changed.
+    changed = {key for key in old_data.keys() & new_data.keys() if old_data[key] != new_data[key]}
+
+    # TODO: evaluate the possibility of losing the diff if some error
+    # happens in the charm before the diff is completely checked (DPE-412).
+    # Convert the new_data to a serializable format and save it for a next diff check.
+    event.relation.data[bucket].update({"data": json.dumps(new_data)})
+
+    # Return the diff with all possible changes.
+    return Diff(added, changed, deleted)
+
+
+class ServiceAccountEvent(RelationEvent):
+    """Base class for Service account events."""
+
+    @property
+    def service_account(self) -> Optional[str]:
+        """Returns the service account was requested."""
+        if not self.relation.app:
+            return None
+
+        return self.relation.data[self.relation.app].get("service-account", "")
+
+    @property
+    def namespace(self) -> Optional[str]:
+        """Returns the namespace that was requested."""
+        if not self.relation.app:
+            return None
+
+        return self.relation.data[self.relation.app].get("namespace", "")
+
+
+class ServiceAccountRequestedEvent(ServiceAccountEvent):
+    """Event emitted when a set of service account/namespace is requested for use on this relation."""
+
+
+class ServiceAccountReleasedEvent(ServiceAccountEvent):
+    """Event emitted when a set of service account/namespace is released."""
+
+
+class IntegrationHubProvidesEvents(CharmEvents):
+    """Event descriptor for events raised by ServiceAccountProvider."""
+
+    account_requested = EventSource(ServiceAccountRequestedEvent)
+    account_released = EventSource(ServiceAccountReleasedEvent)
+
+
+class ServiceAccountGrantedEvent(ServiceAccountEvent):
+    """Event emitted when service account are granted on this relation."""
+
+
+class ServiceAccountGoneEvent(RelationEvent):
+    """Event emitted when service account are removed from this relation."""
+
+
+class IntegrationHubRequiresEvents(ObjectEvents):
+    """Event descriptor for events raised by the Requires."""
+
+    account_granted = EventSource(ServiceAccountGrantedEvent)
+    account_gone = EventSource(ServiceAccountGoneEvent)
+
+
+# Integration Hub Provides and Requires
+
+
+class IntegrationHubProvidesData(ProviderData):
```

Comment:
```suggestion
class IntegrationHubProviderData(ProviderData):
```

Can you fix the spelling similarly to my PR on the DP Libs?